src/RadixDlt.NetworkGateway.PostgresIntegration/Services/RoleAssignmentsMapper.cs

```diff
@@ -0,0 +1,151 @@
+ * or use made or its functionality or any performance or operation of any code or protocol
+ * caused by bugs or programming or logic errors or otherwise);
+ *
+ * A. any offer, purchase, holding, use, sale, exchange or transmission of any
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using RadixDlt.NetworkGateway.Abstractions;
+using RadixDlt.NetworkGateway.PostgresIntegration.Models;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using CoreApiModel = RadixDlt.CoreApiSdk.Model;
+using GatewayModel = RadixDlt.NetworkGateway.GatewayApiSdk.Model;
+
+namespace RadixDlt.NetworkGateway.PostgresIntegration.Services;
+
+internal interface IRoleAssignmentsMapper
+{
+    Dictionary<long, GatewayApiSdk.Model.ComponentEntityRoleAssignments> GetEffectiveRoleAssignments(
+        ICollection<Entity> componentEntities, ICollection<EntityRoleAssignmentsOwnerRoleHistory> ownerRoles, ICollection<EntityRoleAssignmentsEntryHistory> roleAssignments);
+}
+
+internal class RoleAssignmentsMapper : IRoleAssignmentsMapper
+{
+    private readonly IRoleAssignmentsKeyProvider _roleAssignmentsKeyProvider;
+
+    public RoleAssignmentsMapper(IRoleAssignmentsKeyProvider roleAssignmentsKeyProvider)
+    {
+        _roleAssignmentsKeyProvider = roleAssignmentsKeyProvider;
+    }
+
+    public Dictionary<long, GatewayModel.ComponentEntityRoleAssignments> GetEffectiveRoleAssignments(
+        ICollection<Entity> componentEntities, ICollection<EntityRoleAssignmentsOwnerRoleHistory> ownerRoles, ICollection<EntityRoleAssignmentsEntryHistory> roleAssignments)
+    {
+        var fungibleResourceAccessRuleKeys = _roleAssignmentsKeyProvider.GetFungibleResourceKeys();
```

Comment:
Minor, but I'd suggest moving the fungible / nonFungible resource key creation into the relevant if block, for a slight performance improvement to avoid some allocations in the 90% case where it's not needed.

Reply:
I'm not sure about that, that method maps multiple entities, so we'll call that method potentially 30x times if all entities will be fungible resources. It's not a big deal for now since it's all stored in memory and read once, will need rework anyway once we start reading keys from blueprint definition.